File: backend/app/api/routes/competitors.py
"""Competitor monitoring API endpoints."""

import logging

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.models.competitor import Competitor
from app.models.competitor_post import CompetitorPost
from app.models.user import User
from app.schemas.competitor import (
    CompetitorCreate,
    CompetitorList,
    CompetitorPostList,
    CompetitorPostResponse,
    CompetitorResponse,
    SyncResult,
    TrendReport,
    ViralAlert,
)
from app.services.competitor_tracker import CompetitorTracker

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/competitors", tags=["competitors"])


@router.post("/", response_model=CompetitorResponse, status_code=201)
async def add_competitor(
    payload: CompetitorCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Competitor:
    """Add a competitor account to track."""
    competitor = Competitor(
        user_id=current_user.id,
        project_id=payload.project_id,
        platform=payload.platform,
        account_id=payload.account_id,
        account_name=payload.account_name,
        account_url=payload.account_url,
    )
    db.add(competitor)
    await db.flush()
    await db.refresh(competitor)
    return competitor


@router.get("/", response_model=CompetitorList)
async def list_competitors(
    project_id: int | None = Query(None),
    platform: str | None = Query(None),
    skip: int = 0,
    limit: int = 50,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """List user's competitors with last sync time."""
    query = select(Competitor).where(Competitor.user_id == current_user.id)
    count_query = select(func.count()).select_from(Competitor).where(
        Competitor.user_id == current_user.id
    )

    if project_id is not None:
        query = query.where(Competitor.project_id == project_id)
        count_query = count_query.where(Competitor.project_id == project_id)
    if platform:
        query = query.where(Competitor.platform == platform)
        count_query = count_query.where(Competitor.platform == platform)

    query = query.order_by(Competitor.created_at.desc()).offset(skip).limit(limit)

    result = await db.execute(query)
    total_result = await db.execute(count_query)

    return {"items": list(result.scalars().all()), "total": total_result.scalar_one()}


@router.delete("/{competitor_id}", status_code=204)
async def remove_competitor(
    competitor_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Remove a competitor (cascades to posts)."""
    result = await db.execute(
        select(Competitor).where(
            Competitor.id == competitor_id,
            Competitor.user_id == current_user.id,
        )
    )
    competitor = result.scalar_one_or_none()
    if not competitor:
        raise HTTPException(status_code=404, detail="Competitor not found")

    await db.delete(competitor)
    await db.flush()


@router.post("/{competitor_id}/sync", response_model=SyncResult)
async def sync_competitor(
    competitor_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SyncResult:
    """Force sync a specific competitor now."""
    # Verify ownership
    result = await db.execute(
        select(Competitor).where(
            Competitor.id == competitor_id,
            Competitor.user_id == current_user.id,
        )
    )
    if not result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Competitor not found")

    tracker = CompetitorTracker(db)
    return await tracker.sync_competitor(competitor_id)


@router.post("/sync-all", response_model=list[SyncResult])
async def sync_all_competitors(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[SyncResult]:
    """Sync all competitors for the current user."""
    tracker = CompetitorTracker(db)
    return await tracker.sync_all_for_user(current_user.id)


@router.get("/trends", response_model=TrendReport)
async def get_trends(
    project_id: int | None = Query(None),
    period_days: int = Query(7, ge=1, le=90),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TrendReport:
    """Get current lane trend report."""
    tracker = CompetitorTracker(db)
    return await tracker.analyze_trends(
        user_id=current_user.id,
        project_id=project_id,
        period_days=period_days,
    )


@router.get("/viral", response_model=list[ViralAlert])
async def get_viral_posts(
    project_id: int | None = Query(None),
    limit: int = Query(20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[ViralAlert]:
    """Get all viral posts across competitors."""
    query = (
        select(CompetitorPost)
        .join(Competitor)
        .where(
            Competitor.user_id == current_user.id,
            CompetitorPost.is_viral.is_(True),
        )
    )
    if project_id is not None:
        query = query.where(Competitor.project_id == project_id)

    query = query.order_by(
        (CompetitorPost.likes + CompetitorPost.comments + CompetitorPost.shares).desc()
    ).limit(limit)

    result = await db.execute(query)
    posts = list(result.scalars().all())

    # Build alerts with competitor info
    alerts = []
    for post in posts:
        comp_result = await db.execute(
            select(Competitor).where(Competitor.id == post.competitor_id)
        )
        competitor = comp_result.scalar_one_or_none()

        tracker = CompetitorTracker(db)
        avg = await tracker._get_avg_engagement(post.competitor_id)
        engagement = post.likes + post.comments + post.shares
        ratio = engagement / avg if avg > 0 else 0.0

        alerts.append(
            ViralAlert(
                post_id=post.id,
                competitor_name=competitor.account_name if competitor else "Unknown",
                platform=competitor.platform if competitor else "unknown",
                title=post.title,
                likes=post.likes,
                comments=post.comments,
                shares=post.shares,
                views=post.views,
                published_at=post.published_at,
                engagement_ratio=ratio,
            )
        )

    return alerts


@router.get("/{competitor_id}/posts", response_model=CompetitorPostList)
async def get_competitor_posts(
    competitor_id: int,
    skip: int = 0,
    limit: int = 20,
    viral_only: bool = False,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Get tracked posts for a specific competitor."""
    # Verify ownership
    comp_result = await db.execute(
        select(Competitor).where(
            Competitor.id == competitor_id,
            Competitor.user_id == current_user.id,
        )
    )
    if not comp_result.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Competitor not found")

    query = select(CompetitorPost).where(CompetitorPost.competitor_id == competitor_id)
    count_query = select(func.count()).select_from(CompetitorPost).where(
        CompetitorPost.competitor_id == competitor_id
    )

    if viral_only:
        query = query.where(CompetitorPost.is_viral.is_(True))
        count_query = count_query.where(CompetitorPost.is_viral.is_(True))

    query = query.order_by(CompetitorPost.fetched_at.desc()).offset(skip).limit(limit)

    result = await db.execute(query)
    total_result = await db.execute(count_query)

    return {"items": list(result.scalars().all()), "total": total_result.scalar_one()}

File: frontend/src/services/competitors.ts
import { api } from './api'

/* ── Types ── */

export interface Competitor {
  id: number
  user_id: number
  project_id: number | null
  platform: string
  account_id: string
  account_name: string
  account_url: string | null
  last_synced_at: string | null
  total_posts_tracked: number
  created_at: string
}

export interface CompetitorPost {
  id: number
  competitor_id: number
  platform_post_id: string
  title: string | null
  content_text: string | null
  media_urls: string[] | null
  tags: string[] | null
  likes: number
  comments: number
  shares: number
  views: number
  published_at: string | null
  fetched_at: string
  is_viral: boolean
}

export interface TrendingTopic {
  topic: string
  frequency: number
  avg_engagement: number
  example_titles: string[]
}

export interface ViralAlert {
  post_id: number
  competitor_name: string
  platform: string
  title: string | null
  likes: number
  comments: number
  shares: number
  views: number
  published_at: string | null
  engagement_ratio: number
}

export interface TrendReport {
  project_id: number | null
  generated_at: string
  period_days: number
  top_topics: TrendingTopic[]
  viral_posts: ViralAlert[]
  total_posts_analyzed: number
  summary: string
}

export interface SyncResult {
  competitor_id: number
  competitor_name: string
  platform: string
  new_posts: number
  updated_posts: number
  viral_detected: number
  error: string | null
}

interface CompetitorCreatePayload {
  platform: string
  account_id: string
  account_name: string
  account_url?: string
  project_id?: number
}

/* ── API functions ── */

export async function addCompetitor(payload: CompetitorCreatePayload) {
  const { data } = await api.post<Competitor>('/v1/competitors/', payload)
  return data
}

export async function listCompetitors(params?: {
  project_id?: number
  platform?: string
  skip?: number
  limit?: number
}) {
  const { data } = await api.get<{ items: Competitor[]; total: number }>(
    '/v1/competitors/',
    { params },
  )
  return data
}

export async function removeCompetitor(id: number) {
  await api.delete(`/v1/competitors/${id}`)
}

export async function syncCompetitor(id: number) {
  const { data } = await api.post<SyncResult>(`/v1/competitors/${id}/sync`)
  return data
}

export async function syncAllCompetitors() {
  const { data } = await api.post<SyncResult[]>('/v1/competitors/sync-all')
  return data
}

export async function getTrends(params?: {
  project_id?: number
  period_days?: number
}) {
  const { data } = await api.get<TrendReport>('/v1/competitors/trends', {
    params,
  })
  return data
}

export async function getCompetitorPosts(
  competitorId: number,
  params?: { skip?: number; limit?: number; viral_only?: boolean },
) {
  const { data } = await api.get<{ items: CompetitorPost[]; total: number }>(
    `/v1/competitors/${competitorId}/posts`,
    { params },
  )
  return data
}

export async function getViralPosts(params?: {
  project_id?: number
  limit?: number
}) {
  const { data } = await api.get<ViralAlert[]>('/v1/competitors/viral', {
    params,
  })
  return data
}
